RequestService/RequestService.Handlers/PutUpdateJobStatusToAppliedForHandler.cs

```diff
@@ -0,0 +1,126 @@
+﻿using MediatR;
+using System.Threading;
+using System.Threading.Tasks;
+using RequestService.Core.Interfaces.Repositories;
+using RequestService.Core.Services;
+using HelpMyStreet.Contracts.RequestService.Request;
+using HelpMyStreet.Contracts.CommunicationService.Request;
+using HelpMyStreet.Contracts.RequestService.Response;
+using HelpMyStreet.Utils.Enums;
+using System.Linq;
+using System;
+using HelpMyStreet.Contracts.GroupService.Request;
+using System.Collections.Generic;
+
+namespace RequestService.Handlers
+{
+    public class PutUpdateJobStatusToAppliedForHandler : IRequestHandler<PutUpdateJobStatusToAppliedForRequest, PutUpdateJobStatusToAppliedForResponse>
+    {
+        private readonly IRepository _repository;
+        private readonly ICommunicationService _communicationService;
+        private readonly IGroupService _groupService;
+        private readonly IJobService _jobService;
+        private const int ADMIN_USERID = -1;
+
+        public PutUpdateJobStatusToAppliedForHandler(IRepository repository, ICommunicationService communicationService, IGroupService groupService, IJobService jobService)
+        {
+            _repository = repository;
+            _communicationService = communicationService;
+            _groupService = groupService;
+            _jobService = jobService;
+        }
+
+        public async Task<PutUpdateJobStatusToAppliedForResponse> Handle(PutUpdateJobStatusToAppliedForRequest request, CancellationToken cancellationToken)
+        {
+            PutUpdateJobStatusToAppliedForResponse response = new PutUpdateJobStatusToAppliedForResponse()
+            {
+                Outcome = UpdateJobStatusOutcome.Unauthorized
+            };
+
+            bool hasPermission = (request.CreatedByUserID == request.VolunteerUserID || await _jobService.HasPermissionToChangeStatusAsync(request.JobID, request.CreatedByUserID, false, cancellationToken));
+
+            if(!hasPermission)
+            {
+                return response;
+            }
+
+            if (_repository.JobIsInProgressWithSameVolunteerUserId(request.JobID, request.VolunteerUserID) || 
+                (_repository.VolunteerHasAlreadyJobForThisRequestWithThisStatus(request.JobID, request.VolunteerUserID, JobStatuses.AppliedFor)))
```

Comment:
👍 Good shoult to check for both InProgress and AppliedFor.  I think we should maybe check for both statuses in PutUpdateJobStatusToInProgressHandler too.